### PR TITLE
Add multiplicative-expr tests for float and decimal types

### PR DESCRIPTION
**Pull request by gimantha**
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/32462

### CODE REVIEW

**Comment by gimantha** [comment]
I will sync this once https://github.com/ballerina-platform/ballerina-spec/pull/935 merged

File: conformance/lang/expressions/multiplicative-expr/division.balt [hunk 1]
```diff
@@ -0,0 +1,4506 @@
...
+    int a = 0xdef;
+    int b = a / 0x12;
+    io:println(b); // @output 198
+    io:println(a / 18); // @output 198
+    io:println(0Xdef / 18); // @output 198
+}
+
+Test-Case: output
+Description: Test division scenarios for different decimal values.
+Labels: multiplicative-expr, numeric-literal, float-point-literal
```
Comment (lasinicl) [inline conformance/lang/expressions/multiplicative-expr/division.balt:4017]: Shall we use `DecimalFloatingPointNumber` instead of `floating-point-literal`?
Don't need `numeric-literal` label? 
Missing `decimal`, `DecimalTypeSuffix` labels.

File: conformance/lang/expressions/multiplicative-expr/division.balt [hunk 2]
```diff
@@ -0,0 +1,4506 @@
...
+     a = -0.0d;
+     b = 0.0d;
+     c = a / b;
+     io:println(c); // @output NaN
+     io:println(-0.0d / 0.0d); // @output NaN
+}
+
+Test-Case: output
+Description: Test division scenarios for different nillable decimal values.
+Labels: multiplicative-expr, numeric-literal, float-point-literal
```
Comment (lasinicl) [inline conformance/lang/expressions/multiplicative-expr/division.balt:4095]: Shall we use `DecimalFloatingPointNumber` instead of `floating-point-literal`?
Don't need `numeric-literal` label? 
Missing `decimal`, `DecimalTypeSuffix`, `optional-type`, `numeric-nil-lifting`, `unary-minus`, `nil-literal`, `value:toBalString` labels.

File: conformance/lang/expressions/multiplicative-expr/division.balt [hunk 3]
```diff
@@ -0,0 +1,4506 @@
...
+
+     a = ();
+     b = 13.E1521d;
+     c = a / b;
+     io:println(c.toBalString()); // @output ()
+}
+
+Test-Case: output
+Description: Test division scenarios for different float values.
+Labels: multiplicative-expr, numeric-literal, float-point-literal
```
Comment (lasinicl) [inline conformance/lang/expressions/multiplicative-expr/division.balt:4171]: Shall we use `DecimalFloatingPointNumber` instead of `floating-point-literal`?
Don't need `numeric-literal` label? 
Missing `float`, `FloatTypeSuffix`, `unary-minus` labels.

File: conformance/lang/expressions/multiplicative-expr/division.balt [hunk 4]
```diff
@@ -0,0 +1,4506 @@
...
+     a = -0.0f;
+     b = 0.0f;
+     c = a / b;
+     io:println(c); // @output NaN
+     io:println(-0.0f / 0.0f); // @output NaN
+}
+
+Test-Case: output
+Description: Test division scenarios for different nillable float values.
+Labels: multiplicative-expr, numeric-literal, float-point-literal
```
Comment (lasinicl) [inline conformance/lang/expressions/multiplicative-expr/division.balt:4249]: Shall we use `DecimalFloatingPointNumber` instead of `floating-point-literal`?
Don't need `numeric-literal` label?
Missing `float`, `FloatTypeSuffix`, `optional-type`, `numeric-nil-lifting`, `unary-minus`, `nil-literal`, `value:toBalString` labels.

File: conformance/lang/expressions/multiplicative-expr/division.balt [hunk 5]
```diff
@@ -0,0 +1,4506 @@
...
+
+     a = ();
+     b = 13.E1521f;
+     c = a / b;
+     io:println(c.toBalString()); // @output ()
+}
+
+Test-Case: output
+Description: Test division by infinity for different float values.
+Labels: multiplicative-expr, numeric-literal, float-point-literal
```
Comment (lasinicl) [inline conformance/lang/expressions/multiplicative-expr/division.balt:4325]: Shall we use `DecimalFloatingPointNumber` instead of `floating-point-literal`?
Don't need `numeric-literal` label? 
Missing `float`, `FloatTypeSuffix`, `unary-minus`, `float:Infinity`, `float:NaN` labels.

File: conformance/lang/expressions/multiplicative-expr/division.balt [hunk 6]
```diff
@@ -0,0 +1,4506 @@
...
+
+public function main() {
+
```
Comment (lasinicl) [inline conformance/lang/expressions/multiplicative-expr/division.balt:4328]:
```suggestion
```

File: conformance/lang/expressions/multiplicative-expr/division.balt [hunk 7]
```diff
@@ -0,0 +1,4506 @@
...
+     a = float:NaN;
+     b = -float:Infinity;
+     c = a / b;
+     io:println(c); // @output NaN
+     io:println(float:NaN / -float:Infinity); // @output NaN
+}
+
+Test-Case: output
+Description: Test division by infinity for different nillable float values.
+Labels: multiplicative-expr, numeric-literal, float-point-literal
```
Comment (lasinicl) [inline conformance/lang/expressions/multiplicative-expr/division.balt:4362]: Shall we use `DecimalFloatingPointNumber` instead of `floating-point-literal`?
Don't need `numeric-literal` label? 
Missing `float`, `FloatTypeSuffix`, `float:Infinity`, `float:NaN`, `optional-type`, `numeric-nil-lifting`, `unary-minus`, `nil-literal`, `value:toBalString` labels.

File: conformance/lang/expressions/multiplicative-expr/division.balt [hunk 8]
```diff
@@ -0,0 +1,4506 @@
...
+
+     a = ();
+     b = -float:Infinity;
+     c = a / b;
+     io:println(c.toBalString()); // @output ()
+}
+
+Test-Case: output
+Description: Test division by NaN for different float values.
+Labels: multiplicative-expr, numeric-literal, float-point-literal
```
Comment (lasinicl) [inline conformance/lang/expressions/multiplicative-expr/division.balt:4399]: Missing `float`, `FloatTypeSuffix`, `float:Infinity`, `float:NaN`, `unary-minus` labels.

File: conformance/lang/expressions/multiplicative-expr/division.balt [hunk 9]
```diff
@@ -0,0 +1,4506 @@
...
+     a = -float:Infinity;
+     b = float:NaN;
+     c = a / b;
+     io:println(c); // @output NaN
+     io:println(a / b); // @output NaN
+}
+
+Test-Case: output
+Description: Test division by NaN for different nillable float values.
+Labels: multiplicative-expr, numeric-literal, float-point-literal
```
Comment (lasinicl) [inline conformance/lang/expressions/multiplicative-expr/division.balt:4431]: Missing `float`, `FloatTypeSuffix`, `float:Infinity`, `float:NaN`, `optional-type`, `numeric-nil-lifting`, `unary-minus`, `nil-literal`, `value:toBalString` labels.

File: conformance/lang/expressions/multiplicative-expr/division.balt [hunk 10]
```diff
@@ -0,0 +1,4506 @@
...
+
+     a = -float:Infinity;
+     b = float:NaN;
+     c = a / b;
+     io:println(c); // @output NaN
+     io:println(a / b); // @output NaN
+}
+
+Test-Case: output
+Description: Test division by NaN for different nillable float values.
```
Comment (lasinicl) [inline conformance/lang/expressions/multiplicative-expr/division.balt:4430]:
```suggestion
Description: Test division by NaN for different nilable float values.
```

File: conformance/lang/expressions/multiplicative-expr/division.balt [hunk 11]
```diff
@@ -0,0 +1,4506 @@
...
+
+     a = ();
+     b = float:NaN;
+     c = a / b;
+     io:println(c.toBalString()); // @output ()
+}
+
+Test-Case: output
+Description: Test float division scenarios for user-defined subtypes of float.
+Labels: multiplicative-expr, numeric-literal, float-literal, module-type-defn
```
Comment (lasinicl) [inline conformance/lang/expressions/multiplicative-expr/division.balt:4462]: Missing `float`, `FloatTypeSuffix`, `unary-minus`, `unary-plus` labels.

File: conformance/lang/expressions/multiplicative-expr/division.balt [hunk 12]
```diff
@@ -0,0 +1,4506 @@
...
+
+type Floats -2.0f|-1.0f|0.0f|1.0f|2.0f;
+
```
Comment (lasinicl) [inline conformance/lang/expressions/multiplicative-expr/division.balt:4465]: Shall we add some more tests with user-defined subtypes like
```ballerina
const A = 1.0f;
const B = 2.0f;

type Floats1 -2.0f|-1.0f|0.0f|1.0f|2.0f;

type Floats2 A|B;

type Floats3 Floats1|Floats2;
public function main() {
    Floats2 a = 1.0;
    B b = 2.0;
    Floats1|Floats2 c = 2.0f;
    float|Floats1 d = 20f;

    io:println(a/b);
    io:println(a/c);

    io:println(b/c);

    //io:println(a/d); // this seems to be not working
}
```

File: conformance/lang/expressions/multiplicative-expr/division.balt [hunk 13]
```diff
@@ -0,0 +1,4506 @@
...
+    a = -0.0;
+    b = +0.0;
+    Floats e = 2.0f;
+    io:println(e / a); // @output -Infinity
+    io:println(e / b); // @output Infinity
+}
+
+Test-Case: output
+Description: Test decimal division scenarios for user-defined subtypes of decimal.
+Labels: multiplicative-expr, numeric-literal, decimal-literal, module-type-defn
```
Comment (lasinicl) [inline conformance/lang/expressions/multiplicative-expr/division.balt:4489]: Missing `decimal`, `DecimalTypeSuffix`, `unary-minus` labels.

File: conformance/lang/expressions/multiplicative-expr/division.balt [hunk 14]
```diff
@@ -0,0 +1,4506 @@
...
+
+type Decimals -2.0d|-1.0d|0.0d|1.0d|2.0d;
+
```
Comment (lasinicl) [inline conformance/lang/expressions/multiplicative-expr/division.balt:4492]: Shall we add some more tests with user-defined subtypes similar to what was mentioned in the previous comment?

File: conformance/lang/expressions/multiplicative-expr/division.balt [hunk 15]
```diff
@@ -0,0 +1,4506 @@
...
+    decimal d = a / b;
+
+    io:println(a / a); // @output 1.0
+    io:println(c); // @output 1.0
+    io:println(a / b); // @output 0.5
+    io:println(d); // @output 0.5
+
+    io:println(b / a); // @output 2.0
+    io:println(b / b); // @output 1.0
+}
```
Comment (lasinicl) [inline conformance/lang/expressions/multiplicative-expr/division.balt:4506]: Shall we add tests for nil lifted scenarios with user-defined subtypes of float and decimal as well?
And also let's check division at the module level too.

File: conformance/lang/expressions/multiplicative-expr/multiplication.balt [hunk 16]
```diff
@@ -0,0 +1,3989 @@
...
+    int a = 0xdef;
+    int b = a * 0x12;
+    io:println(b); // @output 64206
+    io:println(a * 18); // @output 64206
+    io:println(0Xdef * 18); // @output 64206
+}
+
+Test-Case: output
+Description: Test multiplication scenarios for different decimal values.
+Labels: multiplicative-expr, numeric-literal, float-point-literal
```
Comment (lasinicl) [inline conformance/lang/expressions/multiplicative-expr/multiplication.balt:3450]: Since the multiplication tests also have a similar format as the division, let's address the same suggestions for multiplication as well.

File: conformance/lang/expressions/multiplicative-expr/multiplicative_expr.balt [hunk 17]
```diff
@@ -0,0 +1,418 @@
...
+
+Test-Case: error
+Description: Test the static type of the result being the basic type of the operands via invalid assignment,
+             where operands are of types that are user-defined subtypes of float types.
+Labels: multiplicative-expr, float-type-descriptor, numeric-literal, float-literal
+
+type Floats 1f|2f;
+
+public function main(float a, float b, Floats c) {
+    Floats f1  = a * b; // @error static type of multiplicative-expr with operands of float subtypes is float
```
Comment (lasinicl) [inline conformance/lang/expressions/multiplicative-expr/multiplicative_expr.balt:389]: Shall we update the unused variables with `_`?

File: conformance/lang/expressions/multiplicative-expr/multiplicative_expr.balt [hunk 18]
```diff
@@ -0,0 +1,418 @@
...
+public function main() {
+    int? a1 = 1 * 0x8000000000000000; // @error operands for * with int? expected type should be int
+    int? a2 = -0x8000000000000001 / 2; // @error operands for / with int? expected type should be int
+    int? a3 = 0x8000000000000000 % -0X4FFFFFFFFFFFFFFF6; // @error operands for % with int? expected type should be int
+}
+
+Test-Case: error
+Description: Test the static type of the result being the basic type of the operands via invalid assignment,
+             where operands are of types that are user-defined subtypes of float types.
+Labels: multiplicative-expr, float-type-descriptor, numeric-literal, float-literal
```
Comment (lasinicl) [inline conformance/lang/expressions/multiplicative-expr/multiplicative_expr.balt:384]:
```suggestion
Labels: multiplicative-expr, DecimalFloatingPointNumber, float, FloatTypeSuffix, module-type-defn
```

File: conformance/lang/expressions/multiplicative-expr/multiplicative_expr.balt [hunk 19]
```diff
@@ -0,0 +1,418 @@
...
+
+type Floats 1f|2f;
+
+public function main(float a, float b, Floats c) {
+    Floats f1  = a * b; // @error static type of multiplicative-expr with operands of float subtypes is float
+    Floats f2 = a / b; // @error static type of multiplicative-expr with operands of float subtypes is float
+    Floats f3 = a % b; // @error static type of multiplicative-expr with operands of float subtypes is float
+}
+
+Test-Case: error
```
Comment (lasinicl) [inline conformance/lang/expressions/multiplicative-expr/multiplicative_expr.balt:394]: Same as the previous test?

File: conformance/lang/expressions/multiplicative-expr/multiplicative_expr.balt [hunk 20]
```diff
@@ -0,0 +1,418 @@
...
+Description: Test the static type of the result being the basic type of the operands via invalid assignment,
+             where operands are of types that are user-defined subtypes of decimal types.
+Labels: multiplicative-expr, decimal-type-descriptor, numeric-literal, decimal-literal
```
Comment (lasinicl) [inline conformance/lang/expressions/multiplicative-expr/multiplicative_expr.balt:410]:
```suggestion
Labels: multiplicative-expr, decimal, DecimalFloatingPointNumber, DecimalTypeSuffix, module-type-defn
```

File: conformance/lang/expressions/multiplicative-expr/multiplicative_expr.balt [hunk 21]
```diff
@@ -0,0 +1,418 @@
...
+
+type Decimals 1d|2d;
+
+public function main(decimal a, decimal b, Decimals c) {
+    Decimals f1  = a * b; // @error static type of multiplicative-expr with operands of decimal subtypes is decimal
+    Decimals f2 = a / b; // @error static type of multiplicative-expr with operands of decimal subtypes is decimal
+    Decimals f3 = a % b; // @error static type of multiplicative-expr with operands of decimal subtypes is decimal
+}
```
Comment (lasinicl) [inline conformance/lang/expressions/multiplicative-expr/multiplicative_expr.balt:418]: Shall we add similar tests with nilable types as well?

File: conformance/lang/expressions/multiplicative-expr/remainder.balt [hunk 22]
```diff
@@ -0,0 +1,4348 @@
...
+    int a = 0xdef;
+    int b = a % 0x12;
+    io:println(b); // @output 3
+    io:println(a % 18); // @output 3
+    io:println(0Xdef % 18); // @output 3
+}
+
+Test-Case: output
+Description: Test remainder operation for different float values.
+Labels: multiplicative-expr, numeric-literal, float-point-literal
```
Comment (lasinicl) [inline conformance/lang/expressions/multiplicative-expr/remainder.balt:4033]:
```suggestion
Labels: multiplicative-expr, float, float:NaN, float:Infinity, FloatTypeSuffix, DecimalFloatingPointNumber, unary-minus 
````

File: conformance/lang/expressions/multiplicative-expr/remainder.balt [hunk 23]
```diff
@@ -0,0 +1,4348 @@
...
+     a = -0.0f;
+     b = 0.0f;
+     c = a % b;
+     io:println(c); // @output NaN
+     io:println(-0.0f % 0.0f); // @output NaN
+}
+
+Test-Case: output
+Description: Test remainder operation for different nillable float values.
+Labels: multiplicative-expr, numeric-literal, float-point-literal
```
Comment (lasinicl) [inline conformance/lang/expressions/multiplicative-expr/remainder.balt:4111]:
```suggestion
Labels: multiplicative-expr, float, float:NaN, float:Infinity, FloatTypeSuffix, DecimalFloatingPointNumber, optional-type, numeric-nil-lifting, nil-literal, value:toBalString, unary-minus
````

File: conformance/lang/expressions/multiplicative-expr/remainder.balt [hunk 24]
```diff
@@ -0,0 +1,4348 @@
...
+
+     a = ();
+     b = 5.34f;
+     c = a % b;
+     io:println(c.toBalString()); // @output ()
+}
+
+Test-Case: output
+Description: Test remainder operation for different decimal values.
+Labels: multiplicative-expr, numeric-literal, float-point-literal
```
Comment (lasinicl) [inline conformance/lang/expressions/multiplicative-expr/remainder.balt:4187]:
```suggestion
Labels: multiplicative-expr, decimal, DecimalTypeSuffix, DecimalFloatingPointNumber, unary-minus
````

File: conformance/lang/expressions/multiplicative-expr/remainder.balt [hunk 25]
```diff
@@ -0,0 +1,4348 @@
...
+     a = -0.0d;
+     b = 0.0d;
+     c = a % b;
+     io:println(c); // @output NaN
+     io:println(-0.0d % 0.0d); // @output NaN
+}
+
+Test-Case: output
+Description: Test remainder operation for different nillable decimal values.
+Labels: multiplicative-expr, numeric-literal, float-point-literal
```
Comment (lasinicl) [inline conformance/lang/expressions/multiplicative-expr/remainder.balt:4247]:
```suggestion
Labels: multiplicative-expr, decimal, DecimalTypeSuffix, DecimalFloatingPointNumber, optional-type, numeric-nil-lifting, nil-literal, value:toBalString, unary-minus
````

File: conformance/lang/expressions/multiplicative-expr/remainder.balt [hunk 26]
```diff
@@ -0,0 +1,4348 @@
...
+     io:println(c.toBalString()); // @output ()
+}
+
+
+Test-Case: output
+Description: Test remainder operation for user-defined subtypes of float.
+Labels: multiplicative-expr, numeric-literal, float-literal, module-type-defn
+
+type Floats -2.0f|-1.0f|0.0f|1.0f|2.0f|-0.0f;
+
```
Comment (lasinicl) [inline conformance/lang/expressions/multiplicative-expr/remainder.balt:4312]: Let's add some more tests for remainder with user-defined types of float and decimal similar to what we did for the division. Let's test with nilable user-defined subtypes as well.

**Comment by lasinicl** [comment]
Shall we do the function name changes too?

File: conformance/lang/expressions/multiplicative-expr/division.balt [hunk 27]
```diff
@@ -4033,3 +4033,649 @@ Description: Test integer division panic at module-level.
...
+     io:println(c); // @output NaN
+     io:println(float:NaN / -float:Infinity); // @output NaN
+}
+
+Test-Case: output
+Description: Test division by infinity for different nillable float values.
+Labels: multiplicative-expr, DecimalFloatingPointNumber, float, FloatTypeSuffix, float:Infinity, float:NaN, optional-type, numeric-nil-lifting, unary-minus, nil-literal, value:toBalString
+
+public function init() {
+
```
Comment (lasinicl) [inline conformance/lang/expressions/multiplicative-expr/division.balt:4387]:
```suggestion
```

File: conformance/lang/expressions/multiplicative-expr/division.balt [hunk 28]
```diff
@@ -4033,3 +4033,649 @@ Description: Test integer division panic at module-level.
...
+    io:println(a / a); // @output 1.0
+    io:println(c); // @output 1.0
+    io:println(a / b); // @output 0.5
+    io:println(d); // @output 0.5
+
+    io:println(b / a); // @output 2.0
+    io:println(b / b); // @output 1.0
+}
```
Comment (lasinicl) [inline conformance/lang/expressions/multiplicative-expr/division.balt:4681]: We need to add a test to cover
```
If the contextually expected type for a multiplicative-expr is T, then the contextualy expected type for both operand expressions is T', where a value v is in T' if it belongs to int, decimal or float, and T contains a value with the same basic type as v.
```
in the spec.
```ballerina
public function main() {
    float a = 1 / 2;
    io:println(a);
}
```
```ballerina
public function main() {
    decimal a = 1 / 2;
    io:println(a);
}
```
Let's also add for multiplication and remainder too.

File: conformance/lang/expressions/multiplicative-expr/division.balt [hunk 29]
```diff
@@ -4033,3 +4033,649 @@ Description: Test integer division panic at module-level.
...
+
+    io:println(a / b); // @output 0.5
+    io:println(a / c); // @output 0.5
+
+    io:println(b / c); // @output 1.0
+
+    io:println(a / d); // @output 0.05
+    io:println(d / a); //@output 20.0
+}
+
```
Comment (lasinicl) [inline conformance/lang/expressions/multiplicative-expr/division.balt:4587]: We can also add some tests with different expressions as operands in `multiplicative-expr` (for `*`, `/`,  `%`) with float and decimal as we have done for `multiplicative-expr` with int. 
ex: https://github.com/ballerina-platform/ballerina-spec/blob/conformance-tests/conformance/lang/expressions/multiplicative-expr/division.balt#L3237-L3362

**Comment by hasithaa** [comment]
The Team is busy with finalizing the SwanLake GA release. Hence closing this PR temporarily.

**Comment by SasinduDilshara** [comment]
We need to add test cases for floats and decimal datatypes which are initiate using integers. 

Scenario 1
```
decimal a = 3;
decimal b = 4;
decimal c = a * b;
```

Scenario 2
```
float a = 3;
float b = 4;
float c = a * b;
```

**Comment by gimantha** [comment]
> Don't we need to add test-cases for decimal numbers with too many exponents.
> 
> Scenario
> 
> ```
> decimal a = 1.7976931348623157e+38000000000 * 20;
> io:println(a);
> 
> Error Message - too many exponents found in decimal value '1.7976931348623157e+38000000000'
> ```

We can add, but this is related to decimal type itself. you would get this error even if you don't have a multiplicative expression

**Comment by SasinduDilshara** [comment]
> > Don't we need to add test-cases for decimal numbers with too many exponents.
> > Scenario
> > ```
> > decimal a = 1.7976931348623157e+38000000000 * 20;
> > io:println(a);
> > 
> > Error Message - too many exponents found in decimal value '1.7976931348623157e+38000000000'
> > ```
> 
> We can add, but this is related to decimal type itself. you would get this error even if you don't have a multiplicative expression

Yes. Then It is ok.

**Comment by SasinduDilshara** [comment]
> > Don't we need to add test-cases for decimal numbers with too many exponents.
> > Scenario
> > ```
> > decimal a = 1.7976931348623157e+38000000000 * 20;
> > io:println(a);
> > 
> > Error Message - too many exponents found in decimal value '1.7976931348623157e+38000000000'
> > ```
> 
> We can add, but this is related to decimal type itself. you would get this error even if you don't have a multiplicative expression

Is this reason same for test multiplication of `f` floating numbers and `F` floating numbers separately.
Because there are no test-cases for floating number end with `F`
ex:-
```
float a = 23.84F;
float b = 23.84F;
io:println(a * b);
```

**Comment by gimantha** [comment]
> > > Don't we need to add test-cases for decimal numbers with too many exponents.
> > > Scenario
> > > ```
> > > decimal a = 1.7976931348623157e+38000000000 * 20;
> > > io:println(a);
> > > 
> > > Error Message - too many exponents found in decimal value '1.7976931348623157e+38000000000'
> > > ```
> > 
> > 
> > We can add, but this is related to decimal type itself. you would get this error even if you don't have a multiplicative expression
> 
> Is this reason same for test multiplication of `f` floating numbers and `F` floating numbers separately. Because there are no test-cases for floating number end with `F` ex:-
> 
> ```
> float a = 23.84F;
> float b = 23.84F;
> io:println(a * b);
> ```
Yes. But i think we can add few tests for `f` and `F` cases

File: conformance/lang/expressions/multiplicative-expr/multiplication.balt [hunk 30]
```diff
@@ -3466,3 +3466,654 @@ Description: Test integer multiplication panic at module-level.
...
+
+     a = ();
+     b = 0.0d;
+     c = a * b;
+     io:println(c.toBalString()); // @output ()
+}
+
+Test-Case: output
+Description: Test multiplication scenarios for different float values.
+Labels: multiplicative-expr, DecimalFloatingPointNumber
```
Comment (chiranSachintha) [inline conformance/lang/expressions/multiplicative-expr/multiplication.balt:3626]: Shall we add a few tests with type cast expr

File: conformance/lang/expressions/multiplicative-expr/division.balt [hunk 31]
```diff
@@ -4033,3 +4033,647 @@ Description: Test integer division panic at module-level.
...
+     a = ();
+     b = 13.E1521d;
+     c = a / b;
+     io:println(c.toBalString()); // @output ()
+}
+
+Test-Case: output
+Description: Test division scenarios for different float values.
+Labels: multiplicative-expr, DecimalFloatingPointNumber, float, FloatTypeSuffix, unary-minus
```
Comment (chiranSachintha) [inline conformance/lang/expressions/multiplicative-expr/division.balt:4193]: Shall we add a test with `float:PI`

File: conformance/lang/expressions/multiplicative-expr/remainder.balt [hunk 32]
```diff
@@ -4049,3 +4049,429 @@ Description: Test integer remainder panic at module-level.
...
+     a = -0.0f;
+     b = 0.0f;
+     c = a % b;
+     io:println(c); // @output NaN
+     io:println(-0.0f % 0.0f); // @output NaN
+}
+
+Test-Case: output
+Description: Test remainder operation for different nillable float values.
+Labels: multiplicative-expr, float, float:NaN, float:Infinity, FloatTypeSuffix, DecimalFloatingPointNumber, optional-type, numeric-nil-lifting, nil-literal, value:toBalString, unary-minus
```
Comment (chiranSachintha) [inline conformance/lang/expressions/multiplicative-expr/remainder.balt:4133]: We can add a few more test cases by assigning them to `var`;

File: conformance/lang/expressions/multiplicative-expr/remainder.balt [hunk 33]
```diff
@@ -4049,3 +4049,429 @@ Description: Test integer remainder panic at module-level.
...
+     a = ();
+     b = 5.34d;
+     c = a % b;
+     io:println(c.toBalString()); // @output ()
+}
+
+
+Test-Case: output
+Description: Test remainder operation for user-defined subtypes of float.
+Labels: multiplicative-expr, float-literal, module-type-defn
```
Comment (chiranSachintha) [inline conformance/lang/expressions/multiplicative-expr/remainder.balt:4331]:
```suggestion
Labels: multiplicative-expr, float, float-literal, module-type-defn
```

File: conformance/lang/expressions/multiplicative-expr/remainder.balt [hunk 34]
```diff
@@ -4049,3 +4049,429 @@ Description: Test integer remainder panic at module-level.
...
+    io:println(d); //@output 1.0
+    io:println(a % b); // @output 1.0
+
+    io:println(b % a); // @output 0.0
+    io:println(b % b); // @output 0.0
+}
+
+Test-Case: output
+Description: Test remainder operation for user-defined subtypes of decimal.
+Labels: multiplicative-expr, decimal-literal, module-type-defn
```
Comment (chiranSachintha) [inline conformance/lang/expressions/multiplicative-expr/remainder.balt:4353]:
```suggestion
Labels: multiplicative-expr, decimal, decimal-literal, module-type-defn
```

File: conformance/lang/expressions/multiplicative-expr/remainder.balt [hunk 35]
```diff
@@ -4049,3 +4049,429 @@ Description: Test integer remainder panic at module-level.
...
+    io:println(d); // @output 1.0
+
+    io:println(b % a); // @output 0.0
+    io:println(b % b); // @output 0.0
+}
+
+Test-Case: output
+Description: Test remainder operation for union types of user-defined subtypes of float.
+Fail-Issue: ballerina-platform/ballerina-lang#33265
+Labels: multiplicative-expr, float-literal, module-type-defn, union-type, float, FloatTypeSuffix
```
Comment (chiranSachintha) [inline conformance/lang/expressions/multiplicative-expr/remainder.balt:4375]:
```suggestion
Labels: multiplicative-expr, float-literal, module-type-defn, module-const-decl, union-type, float, FloatTypeSuffix
```

File: conformance/lang/expressions/multiplicative-expr/remainder.balt [hunk 36]
```diff
@@ -4049,3 +4049,429 @@ Description: Test integer remainder panic at module-level.
...
+    io:println(b % c); // @output 0
+
+    io:println(a % d); // @output 1.0
+    io:println(d % a); //@output 0
+}
+
+Test-Case: output
+Description: Test remainder operation for union types of user-defined subtypes of decimal.
+Fail-Issue: ballerina-platform/ballerina-lang#33266
+Labels: multiplicative-expr, decimal-literal, module-type-defn, union-type, decimal, DecimalTypeSuffix
```
Comment (chiranSachintha) [inline conformance/lang/expressions/multiplicative-expr/remainder.balt:4404]:
```suggestion
Labels: multiplicative-expr, decimal-literal, module-type-defn, module-const-decl, union-type, decimal, DecimalTypeSuffix
```

**Comment by gimantha** [comment]
new PR : https://github.com/ballerina-platform/ballerina-spec/pull/1162
